### DISCUSSION TITLE
Issue Tags

### DISCUSSION THREAD

**Discussion by alwinm** [cholla-hydro/cholla#228] [General]
Our current issue tags are:

bug
duplicate
enhancement
help wanted
invalid
question
wontfix

I added tags to many of the closed issues and played with adding a few:

warning
good first issue

This discussion is to solicit suggestions for new tags and how we define/use current tags.

I propose the following tags/definitions
**bug** = serious bugs requiring high priority attention
**warning** = problems like running out of memory, or warnings/errors that appear when given invalid input or environment, which we leave in the record as helpful warnings to future users but close. Another example is Bug + "it works on my machine". My primary goal is to make potentially useful issues easy to find after we close them, and to separate them from one-and-done bugs. What name would make this purpose most obvious? 
**enhancement** = anything where cholla can be made better, but the lack thereof is not a serious bug
**good first issue** = beginner-friendly task (edit: I realized that github detects this as a default issue for new contributors and has features attached)

Possibility: Delete the question issue tag. Questions can be put under discussions.

EDIT: Proposal for using enhancement tag to see what stage a proposed enhancement has reached: 
naked enhancement + open is just a proposal / investigation to figure out if it's something we must do
enhancement + in progress means it is a todo task with consensus (optional + good first issue)
enhancement + wontfix means it's something we investigated and decided not to do
enhancement + closed means we did it

EDIT: Ways to use bug/warning tag 
A bug can become a warning after discovering the user was missing a library, or ran into a hardware limitation. 
A bug or warning can also gain an enhancement tag when the problem is found (?) or should we just leave bugs as bugs, which are implicitly marked as solved when closed?
A bug or warning can also gain "wontfix"

Deprecated ideas post EDIT:

Some other ideas:
**in progress** = a task which has begun so before taking it on, make sure to check the issue comments to see who else is working on it. 
**investigation** = an issue documenting an investigation, for something where it's not obvious if it would be an enhancement or problem or not, and not really a question (maybe belongs more in Discussion)
**proposal** = an issue documenting a proposal (but this probably belongs more in Discussion)

**Comment by bcaddy** [comment]
I think we should have tags for each major physics module and section. I.e. mhd, gravity, boundary conditions, etc. This is probably something we should just add as we go.

I think "Warning" or "FYI" are good labels for that category.

I like "in progress" and I think it covers the "investigation" category as well.

I think "enhancement" and "proposal" are probably the same category as well since a proposal is really just a pre-enhancement and we can always mark it as "wont fix" later. "Enhancement" is the standard so I recommend we stick with that.

What are the special features of "good first issue"? I googled it but didn't find anything useful.

**Reply by alwinm** [reply]
re: good first issue. I do not know either, honestly. I just saw it was something github recommended so I decided to go with the default name rather than "easy" or "beginner friendly". It sounds like maybe they point them out to new contributors or something? IDK.  

Agree: tags for each major physics module as the tag becomes useful, add as we go to avoid clutter.

Thank you for the input on the other labels as well. I think it's a good idea to merge some of the proposed labels to avoid clutter. 

In my mind the difference between in progress and investigation is "in progress" is something agreed to be done and someone is working on it, whereas investigation is something where people are still looking into whether there's a bug or improvement and whether something should be done.

I will agree that there seems to be conceptual overlap between "enhancement" "proposal" "investigation" "in progress." What I wanted to capture is the answer to two questions at a glance: "Was a consensus formed on doing this?" "Is someone working on it?" but maybe that should be entirely within the scope of the project board and left entirely out of tags. I wasn't (and still am not) sure. 

Maybe along the lines of what you suggest:

naked enhancement + open is just a proposal / investigation to figure out if it's something we must do
enhancement + in progress means it is a todo task 
enhancement + wontfix means it's something we investigated and decided not to do
enhancement + closed means we did it

**Reply by bcaddy** [reply]
I like your suggestion for use of the enhancement tag and it's similar to what I've seen in other projects so it won't be a surprise to new people

**Reply by alwinm** [reply]
Thank you for guiding me down this path, I feel much clearer about this way of doing it as well.

**Reply by bcaddy** [reply]
Maybe a "hack day" label? I think this is probably covered by "Good first issue" since anything simple enough for us to do in an afternoon likely is also simple enough for a first issue.

**Reply by alwinm** [reply]
Agreed that it is covered

**Comment by bcaddy** [comment]
As @evaneschneider brought up. We need a label for code cleanup/best practices/standardization tasks. I'm putting some proposals below but I'm not thrilled with any of them. Most have a negative connotation and it would be nice for the connotation to be more positive, others just don't encompass the entire idea IMO.

- 'chore'
- 'code cleanup'
- 'best practices'
- 'technical debt'

**Reply by alwinm** [reply]
I propose some words, trying to be as positive/encompassing as possible:

"polishing"
"cleaning"
"refactoring"
"beautification"

**Reply by bcaddy** [reply]
I like "polishing". It seems to cover all the use cases and doesn't have any other/specific meanings. Maybe "Code Polish" to be more clear?

**Reply by bcaddy** [reply]
@evaneschneider, does "Code Polish" sound good to you for this category?

**Reply by evaneschneider** [reply]
Polish is fine with me. I don't personally think cleanup has a negative connotation, but maybe that's just because I'm a naturally tidy person.. :)

**Reply by evaneschneider** [reply]
Actually, the more I think about it the more I think cleanup is better for some cases, because I think if there's old code lying around that isn't used (my example), removing it really is cleaning, not polishing. Whereas I would associated "polish" with something like the refactor that Bob recently did for HLLD.

**Reply by bcaddy** [reply]
"Code Cleanup/Polish"? Or split it into 2? One for refactoring (called "refactor") and one for cleanup (called "code cleanup")

**Reply by evaneschneider** [reply]
I'm fine with cleanup and refactor.

**Reply by bcaddy** [reply]
Since discussion on this seems to have died down I'm going to add "refactor" and "code cleanup" as tags and assign them to relevant things. We can change it later if we decide we don't like it.

**Reply by alwinm** [reply]
Refactor and code cleanup feel pretty similar to me so I have slight preference for keeping around just one to avoid questions of "is this refactoring or is this cleanup?" but I don't feel strongly about it.

I've slightly edited the label colors to be lighter just because that's my preference but feel free to revert.

**Comment by alwinm** [comment]
I've marked wiki-related issues with a "documentation" tag for now.
